### ISSUE TITLE
mzIdentML Usage Patterns

### ISSUE THREAD

**Issue by mobiusklein** [mobiusklein/psims#2]
The API for `MzIdentMLWriter` is much rougher than `MzMLWriter`, and was designed with the assumption the user would provide generators to lazily provide all of the large collections. This makes it less intuitive to build in complex logic to prepare members of those collections on the fly.

I'm opening this issue to initiate discussion of what would be ideal. @JB-MS, if you have a chance to take a look and let me know how you plan to use the API, we might come to a better design.

**Comment by JB-MS** [comment]
We were planning to integrate the mzid writer as a node in our framework [Ursgal](https://github.com/ursgal/ursgal). At the moment we have .csv as the final output format, but since mzIdentML is the official standard we would like to add a converter to convert the .csv files into mzid, to be compatible with repositories like e.g. PRIDE.

@mobiusklein, if you are interested and have time, you are welcome to join and contribute to Ursgal, and implement your writer into the framework.

I am looking forward to contribute to improvements of the design. At the moment i have no specific adjustment/improvement ideas, but maybe we will see what can be optimized, when it is integrated into Ursgal?

**Comment by mobiusklein** [comment]
I see several nodes in [ursgal](https://github.com/ursgal/ursgal) dealing with csvs. Which one should I be looking at for an equivalent result data structure?

**Comment by JB-MS** [comment]
That is correct. We rely on a workflow right now, which passes the csv from one node to another, which then read and change the csv accordingly.

Simplified example workflow:

mzML --> Search engine --> result file (e.g. xml, csv or mzid) --> conversion to csv --> unify csv (add e.g. missing meta info) --> map peptides --> validate PSMs with e.g. Percolator

See also e.g. this [example script](https://github.com/ursgal/ursgal/blob/master/example_scripts/xtandem_version_comparison_qexactive.py)

In an ideal world we would deal with mzid the whole time, but we think it is also beneficial to have the easily accessible and readable csv format in your workflow.

I imagine the mzid writer at the end of an workflow , after the PSMs were validated using Percolator and then offer a (optional) node to convert these results into a mzid.

What do you think?

**Comment by mobiusklein** [comment]
That's reasonable. My use-case holds intermediary results in a `sqlite3` database before translating the database into either CSV or mzIdentML as needed.

I'm looking for a definition for an output data structure in the ursgal repository, but I don't see a class or consistent `dict`  searching the codebase. Does one exist?

**Comment by JB-MS** [comment]
This is indeed missing at the moment. We rely on mzid to csv conversion using a [Java library ](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3933944/). We used this csv format as a template and modify and add some fields.

I will set up a new parameter in uparams.py (the general parameter storage and required for correct mapping), where the fieldnames for the csv are specified and described. This should be useful to set up a correct converter.
